### PR TITLE
Add SFT support

### LINKED ISSUE #20: Support SFT
# SFT Design for Maester

## Goal
Add minimal supervised fine-tuning (SFT) support to Maester for training on conversation data.

### 1. Data Format
- **Input**: Parquet files with a `conversations` column containing JSON arrays
- **Format**: `[{"role": "user", "content": "..."}, {"role": "assistant", "content": "..."}]`
- **Output**: Dict with `input_ids`, `labels`, `attention_mask` for easy extension

### 2. Architecture
```
maester/sft/
├── messages.py      # Message class with role enum, masking strategies
├── templates.py     # ChatML template (no Jinja2 for now)
├── tokenization.py  # Convert messages → tokens with masking
└── dataset.py       # ConversationParquetDataset extends ParquetDataset
```

### 3. Key Components
- **Message class**: Simple dataclass with `role`, `content`, `masked`, `eot` fields
- **Masking**: Only train on assistant responses by default (labels = -100 for masked tokens)
- **No packing**: One conversation per sequence, right-padded to `max_seq_len` (or minimum of the maximum of the sequence lengths of the samples in the batch and max_seq_len)?
- **Template**: Direct string formatting, not Jinja2 (simpler, no dependencies)

### 4. Integration Points
```python
# In train.py (minimal changes)
if cfg.sft.enabled:
    dataloader = build_sft_data_loader(cfg, rank, world_size)
else:
    dataloader = build_experimental_data_loader(cfg, rank, world_size)
```

### 5. Future Extensions
- **Distillation**: Dict outputs allow adding `teacher_logits` field later
- **Packing**: Can add conversation packing for efficiency
- **Templates**: Easy to add Llama3, Alpaca formats

## Benefits
- Reuses all existing infrastructure (checkpointing, distributed training)
- Not too many lines of code
- No or few modifications to core training loop
- Clear upgrade path for further features

## Definition of Done
- Can load conversation data from Parquet files
- Correctly masks user/system messages in loss computation
- Works with existing distributed training setup
- Checkpointing/resumption works correctly
- Simple example demonstrating usage

## Why Merge This?
1. **Immediate value**: Enables fine-tuning without switching frameworks
2. **Low risk**: Minimal changes to existing code, feature is optional
3. **Foundation for more**: Clean base for distillation, RLHF, and other post-training methods
4. **Maintains philosophy**: Simple, explicit, debuggable - no hidden complexity

### PR DESCRIPTION
**Pull request by rlrs**
Based on Gemma support branch. Closes #20

### CODE REVIEW

**Comment by rlrs** [comment]
There's some annoying code duplication and complexity around the inheritance from ParquetDataset. Not sure how to deal with it except like this for now.

File: maester/config.py [hunk 1]
```diff
@@ -13,34 +13,50 @@
...
     eos_token: int = 128001
     drop_tokens: str = ""
+    # dataset_path: str =  "data/"
+    # datasets: str = "c4_mini,fake_dataset"
+    num_data_workers: int = 1
+    # col_name: str = "tokens"
+    # file_type: str = "arrow"
+
+
+class SFTConfig(BaseSettings):
```
Comment (peter-sk) [inline maester/config.py:29]: this makes a lot of sense

File: maester/config.py [hunk 2]
```diff
@@ -13,34 +13,50 @@
...
+    # Future: distillation settings?
 
 class Config(BaseSettings):
     model_config = SettingsConfigDict(frozen=True, protected_namespaces=(), arbitrary_types_allowed=True, cli_parse_args=False)
 
     # submission/job 
     dump_dir: str = "jobs/"
-    job_name: str = "llama-3.2-3B"
-    num_nodes: int = 32
+    job_name: str = "gemma-sft-test"
```
Comment (peter-sk) [inline maester/config.py:45]: But this doesn't so much. Shouldn't we somehow have two configs instead of overwriting the entire main config.py?

File: maester/models/__init__.py [hunk 3]
```diff
@@ -5,17 +5,28 @@
 # LICENSE file in the root directory of this source tree.
 
 from maester.models.llama import llama2_configs, llama3_configs, mistral_configs, Transformer
+from maester.models.gemma import gemma_configs, gemma3_configs, GemmaTextModel
```
Comment (peter-sk) [inline maester/models/__init__.py:8]: I guess we should first merge gemma3 PR #21  before continuing the review of PR #22.

File: train.py [hunk 4]
```diff
@@ -293,7 +324,16 @@ def loss_fn(pred, labels):
...
                 batch = next(data_iterator)
-                input_ids, labels = batch
+                
+                # Handle different batch formats
+                if cfg.sft is not None:
+                    input_ids = batch["input_ids"]
+                    labels = batch["labels"]
+                    # attention_mask available in batch["attention_mask"] if needed
+                else:
+                    # TODO: update non-sft data loader to return dict format too?
```
Comment (peter-sk) [inline train.py:334]: good idea to always have a dict - the overhead is negligible and it reduces the amount of custom code needed

**Review by peter-sk — CHANGES_REQUESTED** [review]
Let's get gemma3 merged and then merge it with this and continue.

**Comment by rlrs** [comment]
@codex review

File: maester/models/gemma/model.py [hunk 5]
```diff
@@ -559,36 +544,76 @@ def forward(
...
+                    make_document_mask_wrapper(
+                        make_sliding_window_mask_fn(self.config.sliding_window_size),
+                        document_ids,
+                    ),
+                    None,
+                    None,
+                    seq_len,
+                    seq_len,
+                    device=tokens.device,
+                )
```
[the remark this answers is not in the document]
Comment (rlrs) [reply maester/models/gemma/model.py:579]: @codex Write a fix for this (pass in B)

File: maester/models/gemma/model.py [hunk 6]
```diff
@@ -173,8 +179,50 @@ def mask_fn(b, h, q_idx, kv_idx):
...
+        b_flat = b_long.reshape(-1)
+        rows = torch.index_select(doc_ids, 0, b_flat)
+        rows = rows.view(*b_long.shape, seq_len)
+
+        q_docs = torch.gather(rows, -1, q_long.reshape(*q_long.shape, 1)).squeeze(-1)
+        k_docs = torch.gather(rows, -1, kv_long.reshape(*kv_long.shape, 1)).squeeze(-1)
+        same_doc = q_docs == k_docs
+
+        base_mask = base_mask_fn(b_long, h_long, q_long, kv_long)
+        return same_doc & base_mask
```
[the remark this answers is not in the document]
Comment (rlrs) [reply maester/models/gemma/model.py:220]: This is wrong, added a test for it to be sure.

**Comment by rlrs** [comment]
@codex review
